src/Microsoft.AspNetCore.Mvc.Core/Builder/MvcApplicationBuilderExtensions.cs

```diff
@@ -99,5 +92,58 @@ public static IApplicationBuilder UseMvcWithDefaultRoute(this IApplicationBuilde
 
             return app.UseRouter(routes.Build());
         }
+
+        public static IApplicationBuilder UseMvcWithEndpoint(
+            this IApplicationBuilder app,
+            Action<MvcEndpointInfoBuilder> configureRoutes)
+        {
+            if (app == null)
+            {
+                throw new ArgumentNullException(nameof(app));
+            }
+
+            if (configureRoutes == null)
+            {
+                throw new ArgumentNullException(nameof(configureRoutes));
+            }
+
+            VerifyMvcIsRegistered(app);
+
+            var endpointDataSources = app.ApplicationServices.GetRequiredService<IEnumerable<EndpointDataSource>>();
+
+            MvcEndpointDataSource mvcEndpointDataSource = null;
+            foreach (var dataSource in endpointDataSources)
+            {
+                if (dataSource is MvcEndpointDataSource ds)
+                {
+                    mvcEndpointDataSource = ds;
+                    break;
+                }
+            }
+
+            var constraintResolver = app.ApplicationServices.GetRequiredService<IInlineConstraintResolver>();
```

Comment:
can I just say that I hate this thing so much (`IInlineConstraintResolver`)

Reply:
What don't you like?

Reply:
This is the one touch point between *constructable* routing objects and services. It leaks into everything.